File: study_tracker_flutter/lib/data/repositories/ai_challenge_repository.dart
import 'package:drift/drift.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import '../db/database.dart' as db;
import '../db/database_provider.dart';
import '../../domain/domain.dart' as domain;

class AiChallengeRepository {
  final db.AppDatabase _db;

  AiChallengeRepository(this._db);

  Future<int> _activeProfileId() async {
    final row = await (_db.select(_db.appSettings)
          ..where((t) => t.key.equals('currentProfileId')))
        .getSingleOrNull();
    return int.tryParse(row?.value ?? '1') ?? 1;
  }

  Future<List<domain.AiChallenge>> getAll() async {
    final profileId = await _activeProfileId();
    final rows = await (_db.select(_db.aiChallenges)
          ..where((t) => t.profileId.equals(profileId))
          ..orderBy([(t) => OrderingTerm.desc(t.createdAt)]))
        .get();
    return rows.map(_mapRow).toList();
  }

  Stream<List<domain.AiChallenge>> watchAll() {
    final profile = _db.select(_db.appSettings)..where((t) => t.key.equals('currentProfileId'));
    return profile
        .watchSingleOrNull()
        .map((row) => int.tryParse(row?.value ?? '1') ?? 1)
        .distinct()
        .asyncExpand((profileId) {
      final query = _db.select(_db.aiChallenges)
        ..where((t) => t.profileId.equals(profileId))
        ..orderBy([(t) => OrderingTerm.desc(t.createdAt)]);
      return query.watch().map((rows) => rows.map(_mapRow).toList());
    });
  }

  Future<domain.AiChallenge?> getById(String id) async {
    final profileId = await _activeProfileId();
    final row = await (_db.select(_db.aiChallenges)
          ..where((t) => t.id.equals(id) & t.profileId.equals(profileId)))
        .getSingleOrNull();
    if (row == null) return null;
    return _mapRow(row);
  }

  Future<domain.AiChallenge?> getActiveByTier(domain.AiChallengeTier tier) async {
    final profileId = await _activeProfileId();
    final nowIso = DateTime.now().toIso8601String();
    final row = await (_db.select(_db.aiChallenges)
          ..where((t) =>
              t.profileId.equals(profileId) &
              t.tier.equals(tier.name) &
              t.completed.equals(false) &
              t.status.equals('active') &
              t.expiresAt.isBiggerThanValue(nowIso))
          ..orderBy([(t) => OrderingTerm.desc(t.createdAt)])
          ..limit(1))
        .getSingleOrNull();
    if (row == null) return null;
    return _mapRow(row);
  }

  Future<void> create(domain.AiChallenge challenge) async {
    final profileId = await _activeProfileId();
    final trimmedId = challenge.id.trim();
    final baseId = trimmedId.isEmpty
        ? '${challenge.tier.name}_${DateTime.now().millisecondsSinceEpoch}'
        : trimmedId;

    var attempt = 0;
    while (true) {
      final candidateId = attempt == 0
          ? baseId
          : 'p${profileId}_${baseId}_${DateTime.now().microsecondsSinceEpoch}_$attempt';

      try {
        await _db.into(_db.aiChallenges).insert(
              db.AiChallengesCompanion.insert(
                id: candidateId,
                profileId: Value(profileId),
                tier: challenge.tier.name,
                title: challenge.title,
                description: challenge.description,
                icon: challenge.icon,
                metric: challenge.metric.name,
                target: challenge.target,
                expiresAt: challenge.expiresAt.toIso8601String(),
                difficulty: challenge.difficulty.name,
                rewardBadgeName: challenge.rewardBadgeName,
                rewardBadgeIcon: challenge.rewardBadgeIcon,
                completed: Value(challenge.completed),
                rawResponse: Value(challenge.rawResponse),
                subTargetsJson: Value(challenge.subTargets?.toJsonString()),
                unitMinMinutes: Value(challenge.unitMinMinutes),
                status: Value(challenge.status.name),
              ),
            );
        return;
      } catch (e) {
        if (_isAiChallengeIdConflict(e) && attempt < 4) {
          attempt++;
          continue;
        }
        rethrow;
      }
    }
  }

  bool _isAiChallengeIdConflict(Object error) {
    final msg = error.toString().toLowerCase();
    return msg.contains('unique constraint failed') && msg.contains('ai_challenges.id');
  }

  Future<void> markCompleted(String id) async {
    final profileId = await _activeProfileId();
    await (_db.update(_db.aiChallenges)
          ..where((t) => t.id.equals(id) & t.profileId.equals(profileId)))
        .write(
      db.AiChallengesCompanion(
        completed: const Value(true),
        status: const Value('completed'),
        updatedAt: Value(DateTime.now()),
      ),
    );
  }

  Future<domain.AiChallenge?> getSlotChallengeForTier(domain.AiChallengeTier tier) async {
    final profileId = await _activeProfileId();
    final row = await (_db.select(_db.aiChallenges)
          ..where((t) =>
              t.profileId.equals(profileId) &
              t.tier.equals(tier.name) &
              t.completed.equals(false) &
              t.status.equals('active'))
          ..orderBy([(t) => OrderingTerm.desc(t.createdAt)])
          ..limit(1))
        .getSingleOrNull();
    if (row == null) return null;
    return _mapRow(row);
  }

  Future<void> setChallengeStatus(String id, domain.AiChallengeStatus status) async {
    final profileId = await _activeProfileId();
    await (_db.update(_db.aiChallenges)
          ..where((t) => t.id.equals(id) & t.profileId.equals(profileId)))
        .write(
      db.AiChallengesCompanion(
        status: Value(status.name),
        updatedAt: Value(DateTime.now()),
      ),
    );
  }

    /// Marks every still-active row in this tier as replaced (user refresh path).
    Future<void> deleteActiveByTier(domain.AiChallengeTier tier) async {
      final profileId = await _activeProfileId();
      await (_db.update(_db.aiChallenges)
            ..where((t) =>
                t.profileId.equals(profileId) &
                t.tier.equals(tier.name) &
                t.status.equals('active')))
          .write(
        db.AiChallengesCompanion(
          status: const Value('replaced'),
          updatedAt: Value(DateTime.now()),
        ),
      );
    }

  domain.AiChallenge _mapRow(db.AiChallengeRow row) {
    return domain.AiChallenge(
      id: row.id,
      tier: domain.AiChallengeTier.fromDb(row.tier),
      title: row.title,
      description: row.description,
      icon: row.icon,
      metric: domain.AiChallengeMetric.fromDb(row.metric),
      target: row.target,
      expiresAt: DateTime.parse(row.expiresAt),
      difficulty: domain.AiChallengeDifficulty.fromDb(row.difficulty),
      rewardBadgeName: row.rewardBadgeName,
      rewardBadgeIcon: row.rewardBadgeIcon,
      completed: row.completed,
      rawResponse: row.rawResponse,
      createdAt: row.createdAt,
      updatedAt: row.updatedAt,
      subTargets: domain.AiMissionSubTargets.tryParse(row.subTargetsJson),
      unitMinMinutes: row.unitMinMinutes,
      status: domain.AiChallengeStatus.fromDb(row.status),
    );
  }
}

final aiChallengeRepositoryProvider = Provider<AiChallengeRepository>(
  (ref) => AiChallengeRepository(ref.watch(databaseProvider)),
);

File: src/core/data/repositories/aiChallengeRepository.ts
import { getDatabase } from '../database';
import type {
  AiChallenge,
  AiChallengeTier,
  AiChallengeDifficulty,
  AiChallengeMetric,
  AiChallengeStatus,
  AiMissionSubTargets
} from '../../domain';
import { getCurrentProfileId } from './profileRepository';

export const aiChallengeRepository = {
  async getAll(): Promise<AiChallenge[]> {
    const database = await getDatabase();
    await ensureTable(database);
    const profileId = await getCurrentProfileId();
    const rows = await database.select<any[]>(
      'SELECT * FROM ai_challenges WHERE profile_id = ? ORDER BY created_at DESC',
      [profileId]
    );
    return rows.map(mapRowToChallenge);
  },

  async getActiveByTier(tier: AiChallengeTier): Promise<AiChallenge | null> {
    const database = await getDatabase();
    await ensureTable(database);
    const profileId = await getCurrentProfileId();
    const rows = await database.select<any[]>(
      `SELECT * FROM ai_challenges
       WHERE profile_id = ? AND tier = ? AND completed = 0 AND status = 'active'
         AND expires_at > CURRENT_TIMESTAMP
       LIMIT 1`,
      [profileId, tier]
    );
    return rows.length > 0 ? mapRowToChallenge(rows[0]) : null;
  },

  /** Current slot mission even if `expires_at` is in the past (grace UI + manual refresh). */
  async getSlotChallengeForTier(tier: AiChallengeTier): Promise<AiChallenge | null> {
    const database = await getDatabase();
    await ensureTable(database);
    const profileId = await getCurrentProfileId();
    const rows = await database.select<any[]>(
      `SELECT * FROM ai_challenges
       WHERE profile_id = ? AND tier = ? AND completed = 0 AND status = 'active'
       ORDER BY datetime(created_at) DESC
       LIMIT 1`,
      [profileId, tier]
    );
    return rows.length > 0 ? mapRowToChallenge(rows[0]) : null;
  },

  async setChallengeStatus(id: string, status: AiChallengeStatus): Promise<void> {
    const database = await getDatabase();
    await ensureTable(database);
    const profileId = await getCurrentProfileId();
    await database.execute(
      `UPDATE ai_challenges SET status = ?, updated_at = CURRENT_TIMESTAMP
       WHERE id = ? AND profile_id = ?`,
      [status, id, profileId]
    );
  },

  async create(challenge: Partial<AiChallenge>): Promise<void> {
    const database = await getDatabase();
    await ensureTable(database);
    const profileId = await getCurrentProfileId();

    if (
      !challenge.id || !challenge.tier || !challenge.title || !challenge.description ||
      !challenge.icon || !challenge.metric || challenge.target === undefined ||
      !challenge.expiresAt || !challenge.difficulty || !challenge.rewardBadgeName ||
      !challenge.rewardBadgeIcon
    ) {
      throw new Error('Challenge payload is incomplete.');
    }

    const status: AiChallengeStatus = challenge.status ?? 'active';
    const subJson =
      challenge.subTargets != null ? JSON.stringify(challenge.subTargets) : null;
    const unitMin =
      challenge.unitMinMinutes !== undefined && challenge.unitMinMinutes !== null
        ? challenge.unitMinMinutes
        : null;

    await database.execute(
      `INSERT INTO ai_challenges (
        id, profile_id, tier, title, description, icon, metric, target,
        expires_at, difficulty, reward_badge_name, reward_badge_icon,
        raw_response, sub_targets_json, unit_min_minutes, status
      ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
      [
        challenge.id,
        profileId,
        challenge.tier,
        challenge.title,
        challenge.description,
        challenge.icon,
        challenge.metric,
        challenge.target,
        challenge.expiresAt,
        challenge.difficulty,
        challenge.rewardBadgeName,
        challenge.rewardBadgeIcon,
        challenge.rawResponse ?? null,
        subJson,
        unitMin,
        status
      ]
    );
  },

  async markCompleted(id: string): Promise<void> {
    const database = await getDatabase();
    await ensureTable(database);
    const profileId = await getCurrentProfileId();
    await database.execute(
      'UPDATE ai_challenges SET completed = 1, status = \'completed\', updated_at = CURRENT_TIMESTAMP WHERE id = ? AND profile_id = ?',
      [id, profileId]
    );
  },

  async deleteActiveByTier(tier: AiChallengeTier): Promise<void> {
    const database = await getDatabase();
    await ensureTable(database);
    const profileId = await getCurrentProfileId();
    await database.execute(
      `UPDATE ai_challenges SET status = 'replaced', updated_at = CURRENT_TIMESTAMP
       WHERE profile_id = ? AND tier = ? AND status = 'active'`,
      [profileId, tier]
    );
  },

  async deleteExpired(): Promise<void> {
    const database = await getDatabase();
    await ensureTable(database);
    const profileId = await getCurrentProfileId();
    await database.execute(
      'DELETE FROM ai_challenges WHERE profile_id = ? AND expires_at <= CURRENT_TIMESTAMP AND completed = 0',
      [profileId]
    );
  }
};

async function ensureTable(database: Awaited<ReturnType<typeof getDatabase>>): Promise<void> {
  await database.execute(
    `CREATE TABLE IF NOT EXISTS ai_challenges (
      id TEXT PRIMARY KEY,
      profile_id INTEGER NOT NULL DEFAULT 1,
      tier TEXT NOT NULL CHECK(tier IN ('daily', 'weekly', 'monthly', 'surprise')),
      title TEXT NOT NULL,
      description TEXT NOT NULL,
      icon TEXT NOT NULL,
      metric TEXT NOT NULL,
      target INTEGER NOT NULL,
      expires_at TEXT NOT NULL,
      difficulty TEXT NOT NULL,
      reward_badge_name TEXT NOT NULL,
      reward_badge_icon TEXT NOT NULL,
      completed INTEGER NOT NULL DEFAULT 0,
      raw_response TEXT,
      created_at TEXT NOT NULL DEFAULT CURRENT_TIMESTAMP,
      updated_at TEXT NOT NULL DEFAULT CURRENT_TIMESTAMP
    )`
  );

  await database.execute('CREATE INDEX IF NOT EXISTS idx_ai_challenges_tier ON ai_challenges(tier)');
  await database.execute('CREATE INDEX IF NOT EXISTS idx_ai_challenges_profile_id ON ai_challenges(profile_id)');
  await database.execute('CREATE INDEX IF NOT EXISTS idx_ai_challenges_completed ON ai_challenges(completed)');
  await database.execute('CREATE INDEX IF NOT EXISTS idx_ai_challenges_expires_at ON ai_challenges(expires_at)');
}

function mapRowToChallenge(row: any): AiChallenge {
  const statusRaw = row.status as string | null | undefined;
  const status: AiChallengeStatus =
    statusRaw === 'completed' || statusRaw === 'expired' || statusRaw === 'replaced' || statusRaw === 'active'
      ? statusRaw
      : 'active';

  let subTargets: AiMissionSubTargets | null | undefined;
  const st = row.sub_targets_json;
  if (typeof st === 'string' && st.trim().length > 0) {
    try {
      const parsed = JSON.parse(st) as AiMissionSubTargets;
      if (parsed && typeof parsed.mode === 'string' && typeof parsed.count === 'number') {
        subTargets = parsed;
      }
    } catch {
      subTargets = null;
    }
  }

  return {
    id: row.id,
    tier: row.tier as AiChallengeTier,
    title: row.title,
    description: row.description,
    icon: row.icon,
    metric: row.metric as AiChallengeMetric,
    target: row.target,
    expiresAt: row.expires_at,
    difficulty: row.difficulty as AiChallengeDifficulty,
    rewardBadgeName: row.reward_badge_name,
    rewardBadgeIcon: row.reward_badge_icon,
    completed: Boolean(row.completed),
    rawResponse: row.raw_response,
    createdAt: row.created_at,
    updatedAt: row.updated_at,
    subTargets: subTargets ?? null,
    unitMinMinutes: row.unit_min_minutes != null ? Number(row.unit_min_minutes) : null,
    status
  };
}
